lib/base/journaldlogger.cpp

```diff
@@ -0,0 +1,79 @@
+#if !defined(_WIN32) && defined(HAVE_SYSTEMD)
+#include "base/journaldlogger.hpp"
+#include "base/journaldlogger-ti.cpp"
+#include "base/configtype.hpp"
+#include "base/statsfunction.hpp"
+#include "base/sysloglogger.hpp"
+#include <systemd/sd-journal.h>
+
+using namespace icinga;
+
+REGISTER_TYPE(JournaldLogger);
+
+REGISTER_STATSFUNCTION(JournaldLogger, &JournaldLogger::StatsFunc);
+
+void JournaldLogger::StatsFunc(const Dictionary::Ptr& status, const Array::Ptr&)
+{
+	DictionaryData nodes;
+
+	for (const JournaldLogger::Ptr& journaldlogger : ConfigType::GetObjectsByType<JournaldLogger>()) {
+		nodes.emplace_back(journaldlogger->GetName(), 1); //add more stats
+	}
+
+	status->Set("journaldlogger", new Dictionary(std::move(nodes)));
+}
+
+void JournaldLogger::OnConfigLoaded()
+{
+	ObjectImpl<JournaldLogger>::OnConfigLoaded();
+
+	String facilityString = GetFacility();
+	m_Facility = SyslogFacilities::toNumber(facilityString);
+}
+
+void JournaldLogger::ValidateFacility(const Lazy<String>& lvalue, const ValidationUtils& utils)
+{
+	ObjectImpl<JournaldLogger>::ValidateFacility(lvalue, utils);
+	if (!SyslogFacilities::validate(lvalue()))
+		BOOST_THROW_EXCEPTION(ValidationError(this, { "facility" }, "Invalid facility specified."));
+}
+
+/**
+ * Processes a log entry and outputs it to journald.
+ *
+ * @param entry The log entry.
+ */
+void JournaldLogger::ProcessLogEntry(const LogEntry& entry)
+{
+	int severity;
+	switch (entry.Severity) {
+		case LogDebug:
+			severity = LOG_DEBUG;
+			break;
+		case LogNotice:
+			severity = LOG_NOTICE;
+			break;
+		case LogWarning:
+			severity = LOG_WARNING;
+			break;
+		case LogCritical:
+			severity = LOG_CRIT;
+			break;
+		case LogInformation:
+		default:
+			severity = LOG_INFO;
+			break;
+	}
+
+	sd_journal_send("MESSAGE=%s", entry.Message.CStr(),
```

Comment:
Something I'm not entirely sure about: would it make sense to add our facility to the message nonetheless? I mean sure, it's there in the structured attributes, but maybe it would still be useful when just looking at `journalctl -u icinga2`? I don't know if all our messages are self-contained enough that they don't require that additional context information. Any second opinion on this, @N-o-X?

Reply:
> would it make sense to add our facility to the message nonetheless?

My 2 cents: No strong opinion. I'd probably prefer short messages as they're now, without redundant information. It would probably be more important for the `SyslogLogger`, where our facility is currently also not inlined.

Reply:
I see. If it's not there for  syslog and nobody complained over the years, it's probably fine.